### PR TITLE
AMDGPU: Add a subtarget feature for fine-grained remote memory support

### PR DESCRIPTION
**Pull request by arsenm**
Atomic access to fine-grained remote memory does not work on all
subtargets. Add a feature for targets where this is expected to work.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96442?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#96443** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96443?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96442** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96442?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#95930** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95930?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95929** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95929?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by AlexVlx** [comment]
Hmm, I might need to reparse this but I'm not entirely sure how this is expected to work in general since PCI-E attachment is not always / entirely knowable at compile time. Have I missed some preamble discussion here?

**Comment by arsenm** [comment]
> Hmm, I might need to reparse this but I'm not entirely sure how this is expected to work in general since PCI-E attachment is not always / entirely knowable at compile time. Have I missed some preamble discussion here?

We do statically know for some of the targets (mostly gfx12 and gfx940) that it's supposed to work. This is the "scope downgrade" vs. "nop" cases in the atomic support table

**Comment by rampitec** [comment]
> We do statically know for some of the targets (mostly gfx12 and gfx940) that it's supposed to work. This is the "scope downgrade" vs. "nop" cases in the atomic support table

Actually not, we do not know the bus. Moreover, we know this is opposite.

**Comment by arsenm** [comment]
> Actually not, we do not know the bus. Moreover, we know this is opposite.

On gfx940/gfx12, we don't need to know the bus to handle the agent scope case correctly. The instructions still function, just always at device scope. We do need to know the bus and/or location to handle the system scope correctly, which will come from the new metadata.

**Comment by yxsamliu** [comment]
If a sub target does not have this feature, does none of the atomic instructions work for fine-grained remote memory, including integer atomic add/xchg/cmpxchg?

**Comment by arsenm** [comment]
No, the string description mentions this is for the non-PCIe supported cases

**Comment by yxsamliu** [comment]
need some tests

**Comment by arsenm** [comment]
> need some tests

This does nothing as it is. The real use patches have thousands of tests

File: llvm/lib/Target/AMDGPU/AMDGPU.td [hunk 1]
```diff
@@ -788,6 +788,14 @@ def FeatureFlatAtomicFaddF32Inst
   "Has flat_atomic_add_f32 instruction"
 >;
 
+def FeatureAgentScopeFineGrainedRemoteMemoryAtomics
+  : SubtargetFeature<"agent-scope-fine-grained-remote-memory-atomics",
+  "HasAgentScopeFineGrainedRemoteMemoryAtomics",
+  "true",
+  "Agent (device) scoped atomic operations not directly supported by "
```
Comment (yxsamliu) [inline llvm/lib/Target/AMDGPU/AMDGPU.td:795]: I feel the description is a little bit confusing, at least for me.

how about

"Agent (device) scoped atomic operations, excluding those directly supported by PCIe (i.e., integer atomic add, exchange, and compare-and-swap), are functional for allocations in host or peer PCIe device memory."

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPU.td]: I dropped the last PCIe reference since XGMI should also work

**Review by arsenm — COMMENTED** [review]
ping

File: llvm/include/llvm/CodeGen/MachineBranchProbabilityInfo.h [hunk 2]
```diff
@@ -14,13 +14,14 @@
...
 
 #include "llvm/CodeGen/MachineBasicBlock.h"
-#include "llvm/CodeGen/MachinePassManager.h"
 #include "llvm/Pass.h"
 #include "llvm/Support/BranchProbability.h"
 
 namespace llvm {
 
-class MachineBranchProbabilityInfo {
+class MachineBranchProbabilityInfo : public ImmutablePass {
```
Comment (jayfoad) [inline llvm/include/llvm/CodeGen/MachineBranchProbabilityInfo.h:22]: Why does the PR include all this unrelated stuff? Which part am I supposed to review? Normally I just look at the "Files changed" tab.

Comment (arsenm) [reply llvm/include/llvm/CodeGen/MachineBranchProbabilityInfo.h:22]: The parts I actually committed. I don't know why "[martinboehme](https://github.com/martinboehme) and others added 14 commits [12 hours ago](https://github.com/llvm/llvm-project/pull/96442#commits-pushed-803aa88)" happened, adding all this other stuff to my personal branch. This has happened several other times, but not consistently

Comment (arsenm) [reply llvm/include/llvm/CodeGen/MachineBranchProbabilityInfo.h:22]: This isn't happening for the other patches further up the stack, just this one for some reason

Comment (arsenm) [reply llvm/include/llvm/CodeGen/MachineBranchProbabilityInfo.h]: I don't know how it ended up like this but I've gotten rid of these commits

File: llvm/lib/Target/AMDGPU/AMDGPU.td [hunk 3]
```diff
@@ -788,6 +788,16 @@ def FeatureFlatAtomicFaddF32Inst
   "Has flat_atomic_add_f32 instruction"
 >;
 
+def FeatureAgentScopeFineGrainedRemoteMemoryAtomics
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPU.td:791]: I originally went with device and changed it to agent to match the syncscope spelling. This was the HSA name

**Comment by arsenm** [comment]
ping

**Review by yxsamliu — APPROVED** [review]

**Comment by arsenm** [comment]
### Merge activity

* **Jul 10, 8:37 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/96442).
